Allow modules to handle RDB loading errors.

oranagra opened the pull request:
This is especially needed in diskless loading, were a short read could have
caused redis to exit. now the module can handle the error and return to the
caller gracefully.

this fixes #5326

antirez commented:
@oranagra thanks Oran, does it also means that modules not handling I/O errors will cause the replica with diskless loading enabled to abort Redis?

oranagra commented:
yes.. that's what the old code does... with this PR, at least modules that wanna play nice have an option to do so...

antirez commented:
@oranagra do I understand correctly that we can't handle the error in place of the module, because it may have half-loaded stuff in memory, and in that case it would leak or alike? That's my first guess, because otherwise we could, like, handle I/O errors in this way.

oranagra commented:
yes, the module must get the error, do some cleanups and abort the loading, and propagate the error upwards (eventually to redis).
later when the loading fails, and redis does decrRefcount on that object, the module should be able to release what it already contains and avoid crashing on that non-normal mt->free call.

antirez commented:
Hello @oranagra, I suggest the following changes to get this merged (as discussed in the chat):

1. Modules that handle I/O errors should mark themselves as such in the onLoad() method. I think we could add an API like:

    RedisModule_DeclareCapability(ctx,REDISMODULE_CAPA_HANDLE_IO_ERRORS);

This should be a bitfield in the module structure I guess.
But if you have a better idea, just tell me.

2. Redis should check if there are modules that export types but don't advertise such capability. In such case diskless loading should be avoided. Such check should be made in three different places AFAIK: at startup, after every MODULE LOAD call, every time CONFIG SET will enable diskless loading. When such conflict is found (diskless loading is on but we can't use it because of the module) Redis should log a message.

3. The module APIs should be documented to explain that module authors generally don't need to handle the I/O errors, but this is only useful in case they want to support diskless loading in replicas, and in such case, explain how such I/O errors can be handled.

4. We should make sure that the error could be checked in a more relaxed way after doing multiple I/O calls probably. Not sure if this is really viable, but would be a plus. So author modules that want to adapt the module to support such capability, can do it more easily by handling the error at the end. However if there is an error condition, I'm not sure how the API calls return value change. For instance what will be the return value of loading a string when there is an I/O error that previously made Redis aborting?

We can zoom a bit on "4" but otherwise I'm pretty convinced 1,2,3 are sane changes.

antirez commented:
About "4", I checked better the PR code, and it looks like anyway users will be get NULL in certain situations and alike, so anyway they should check for errors. The alternative of returning an "I/O error" string does not look much saner just for the sake of doing error handling only later. I guess that at least people can do things like that:

    a = RedisModule_LoadUnsigned(...);
    b = RedisModule_LoadUnsigned(...);
    c = RedisModule_LoadUnsigned(...);
    /* Do I/O check just here. */

It's already an improvement.

antirez commented:
@oranagra I tried to ping you on Slack about that, but is right now unavailable (down) from Italy.

https://status.slack.com/2019-07/b31a6510b8caabcc

oranagra commented:
@antirez thanks for these notes.

1. i suggest a more generic name like RedisModule_SetModuleOptions. so we can use it for other bit flags which are not capabilities. for one, we found it useful to disable the the implicit signalModifiedKey in RM_CloseKey, and add an api that calls it explicitly, so this SetModuleOptions api can set that bit too.

2. if we don't want to return an error in the config set (which we can't since a module can be loaded later), then why do we need to check in 3 different places? isn't it enough to run that check in readSyncBulkPayload before deciding which mode we use?

3. i'll do my best with the documentation.

4. it seems the current state is already that modules can do as many read calls they want, and just check for errors before using this data (just like we did in rdb.c recently)

antirez commented:
@oranagra you are right, I had the feeling that doing the check in readSyncBulkPayload() would be slow but actually does not make sense, is a trivial O(N) operation and N (number of modules) is super small in all the cases. Agreed on all the other things as well, including the API name.

oranagra commented:
@antirez i implemented your comments, merged the current `unstable` (where we already have a test for module rdb loading), and added a test that checks short reads on diskless. (each in a different commit)

antirez commented:
@oranagra thank you, changes merged. Appreciate the good code and the tests.
